cpp/src/cylon/net/ops/gather.cpp

```diff
@@ -153,6 +154,67 @@ cylon::Status cylon::mpi::Gather(const std::shared_ptr<cylon::TableSerializer> &
   return cylon::Status::OK();
 }
 
+cylon::Status cylon::mpi::GatherArrowBuffer(const std::shared_ptr<arrow::Buffer> &buf,
+                                            int gather_root,
+                                            const std::shared_ptr<cylon::CylonContext> &ctx,
+                                            std::vector<std::shared_ptr<arrow::Buffer>> &buffers) {
+
+  std::vector<int32_t> all_buffer_sizes;
+  if (AmIRoot(gather_root, ctx)) {
+    all_buffer_sizes.resize(ctx->GetWorldSize(), 0);
+  }
+
+  int32_t size = static_cast<int32_t>(buf->size());
+  int status = MPI_Gather(&size,
+                          1,
+                          MPI_INT32_T,
+                          all_buffer_sizes.data(),
+                          1,
+                          MPI_INT32_T,
+                          gather_root,
+                          MPI_COMM_WORLD);
+  if (status != MPI_SUCCESS) {
+    return cylon::Status(cylon::Code::ExecutionError, "MPI_Gather failed when receiving buffer sizes!");
+  }
+
+  std::shared_ptr<arrow::Buffer> all_buf = std::move(arrow::AllocateBuffer(0).MoveValueUnsafe());
+  std::vector<int32_t> disps;
+
+  if (AmIRoot(gather_root, ctx)) {
+    auto total_size = std::accumulate(all_buffer_sizes.begin(), all_buffer_sizes.end(), 0);
+    all_buf = std::move(arrow::AllocateBuffer(total_size).MoveValueUnsafe());
+
+    disps.resize(ctx->GetWorldSize(), 0);
+    for (int i = 1; i < all_buffer_sizes.size(); ++i) {
+      disps[i] = disps[i-1] + all_buffer_sizes[i-1];
+    }
+  }
+
+  status = MPI_Gatherv(buf->data(),
+                       size,
+                       MPI_UINT8_T,
+                       (void *)all_buf->data(),
+                       all_buffer_sizes.data(),
+                       disps.data(),
+                       MPI_UINT8_T,
+                       gather_root,
+                       MPI_COMM_WORLD);
+  if (status != MPI_SUCCESS) {
+    return cylon::Status(cylon::Code::ExecutionError, "MPI_Gatherv failed when receiving buffers!");
+  }
+
+  if (gather_root == ctx->GetRank()) {
+    buffers.resize(ctx->GetWorldSize());
+    int32_t buf_start = 0;
+    for (int i = 0; i < ctx->GetWorldSize(); ++i) {
+      buffers[i] = arrow::SliceBuffer(all_buf, buf_start, all_buffer_sizes[i]);
+      buf_start += all_buffer_sizes[i];
+    }
```

Comment:
this calculation is already done I think :slightly_smiling_face: and let's use `arrow::SliceMutableBuffer` instead. I've had some problems previously with `SliceBuffer` because it creates a immutable buffer. 
```suggestion
    buffers.reserve(ctx->GetWorldSize());
    for (int i = 0; i < ctx->GetWorldSize(); ++i) {
      buffers.push_back(arrow::SliceMutableBuffer(all_buf, disps[i], all_buffer_sizes[i]));
    }
```

Reply:
Done